electrum/json_db.py

```diff
@@ -268,17 +347,32 @@ def _convert_value(self, path, key, v):
                 v = constructor(v)
         return v
 
-    def write(self, storage: 'WalletStorage'):
-        with self.lock:
-            self._write(storage)
+    @locked
+    def write(self):
+        if self.storage.file_exists() and not self.storage.is_encrypted():
+            self._append_pending_changes()
+        else:
+            self._write()
+
+    @locked
+    def _append_pending_changes(self):
+        if threading.currentThread().isDaemon():
+            self.logger.warning('daemon thread cannot write db')
+            return
+        if not self.pending_changes:
+            return
+        s = ''.join([',\n' + x for x in self.pending_changes])
```

Comment:
Rather than prepending an integer to every line, it would be cleaner to write that integer in the update.
For example:

```diff --git a/electrum/json_db.py b/electrum/json_db.py
index 579a7f7b4..183754e1d 100644
--- a/electrum/json_db.py
+++ b/electrum/json_db.py
@@ -241,9 +241,12 @@ class JsonDB(Logger):
         if patches:
             # apply patches
             self.logger.info('found %d patches'%len(patches))
+            for n, patch in enumerate(patches):
+                assert patch['id'] == n
             patch = jsonpatch.JsonPatch(patches)
             data = patch.apply(data)
             self.set_modified(True)
+        self.num_patches = len(patches)
         return data, bool(patches)
 
     def maybe_load_incomplete_data(self, s):
@@ -274,6 +277,8 @@ class JsonDB(Logger):
 
     @locked
     def add_patch(self, patch):
+        patch['id'] = self.num_patches
+        self.num_patches += 1
         self.pending_changes.append(json.dumps(patch, cls=self.encoder))
         self.set_modified(True)
 
@@ -392,3 +397,4 @@ class JsonDB(Logger):
         self.storage.write(json_str)
         self.pending_changes = []
         self.set_modified(False)
+        self.num_patches = 0

```

I am not sure if this is really useful, though. The current code already checks that no other process (or `WalletDB` instance in the same process) appended to the file, by checking the file length every time it appends. It seems to me that this is a stronger protection, because the check is performed earlier (at write time, not at consolidation time).

Anyway, I wrote the above patch because I wanted to compare both approaches.